internal/ls/autoimport/registry.go

```diff
@@ -0,0 +1,1030 @@
+		}
+		logger.Logf("Built %d indexes in %v", len(tasks), time.Since(start))
+	}
+}
+
+type failedAmbientModuleLookupSource struct {
+	mu          sync.Mutex
+	fileName    string
+	packageName string
+}
+
+type bucketBuildResult struct {
+	bucket *RegistryBucket
+	// File path to filename and package name
+	possibleFailedAmbientModuleLookupSources *collections.SyncMap[tspath.Path, *failedAmbientModuleLookupSource]
+	// Likely ambient module name
+	possibleFailedAmbientModuleLookupTargets *collections.SyncSet[string]
+}
+
+func (b *registryBuilder) buildProjectBucket(
+	ctx context.Context,
+	projectPath tspath.Path,
+	resolvedPackageNames *collections.Set[string],
+	nodeModulesContainsDependency func(nodeModulesDir tspath.Path, packageName string) bool,
+	logger *logging.LogTree,
+) (*bucketBuildResult, error) {
+	if ctx.Err() != nil {
+		return nil, ctx.Err()
+	}
+
+	start := time.Now()
+	var mu sync.Mutex
+	fileExcludePatterns := b.userPreferences.ParsedAutoImportFileExcludePatterns(b.host.FS().UseCaseSensitiveFileNames())
+	result := &bucketBuildResult{bucket: &RegistryBucket{}}
+	program := b.host.GetProgramForProject(projectPath)
+	getChecker, closePool, checkerCount := createCheckerPool(program)
+	defer closePool()
+	exports := make(map[tspath.Path][]*Export)
+	var wg sync.WaitGroup
+	var ignoredPackageNames collections.Set[string]
+	var skippedFileCount int
+	var combinedStats extractorStats
+
+outer:
+	for _, file := range program.GetSourceFiles() {
+		if program.IsSourceFileDefaultLibrary(file.Path()) {
+			continue
+		}
+		for _, excludePattern := range fileExcludePatterns {
+			if matched, _ := excludePattern.MatchString(file.FileName()); matched {
+				skippedFileCount++
+				continue outer
+			}
+		}
+		if packageName := modulespecifiers.GetPackageNameFromDirectory(file.FileName()); packageName != "" {
+			// Only process this file if it is not going to be processed as part of a node_modules bucket
+			// *and* if it was imported directly (not transitively) by a project file (i.e., this is part
+			// of a package not listed in package.json, but imported anyway).
+			pathComponents := tspath.GetPathComponents(string(file.Path()), "")
+			nodeModulesDir := tspath.GetPathFromPathComponents(pathComponents[:slices.Index(pathComponents, "node_modules")])
+			if nodeModulesContainsDependency(tspath.Path(nodeModulesDir), packageName) {
+				continue
+			}
+			if !resolvedPackageNames.Has(packageName) {
+				ignoredPackageNames.Add(packageName)
+				continue
+			}
+		}
+		wg.Go(func() {
+			if ctx.Err() == nil {
+				checker, done := getChecker()
+				defer done()
+				extractor := b.newExportExtractor("", "", checker)
+				fileExports := extractor.extractFromFile(file)
+				mu.Lock()
+				exports[file.Path()] = fileExports
+				mu.Unlock()
+				stats := extractor.Stats()
+				atomic.AddInt32(&combinedStats.exports, stats.exports)
+				atomic.AddInt32(&combinedStats.usedChecker, stats.usedChecker)
```

Comment:
These can be `atomic.Int32` on the fields themselves (unless you're relying on a struct copy being """legal""").